File: examples/train_shakespeare.cpp
#include <cmath>
#include <algorithm>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <random>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

#include "tiramisu/autograd/grad_mode.hpp"
#include "tiramisu/autograd/ops.hpp"
#include "tiramisu/core/cuda_common.hpp"
#include "tiramisu/core/cuda_memory.hpp"
#include "tiramisu/core/tensor.hpp"
#include "tiramisu/nn/gpt.hpp"
#include "tiramisu/nn/loss.hpp"
#include "tiramisu/optim/adamw.hpp"
#include "tiramisu/optim/grad_clip.hpp"
#include "tiramisu/optim/lr_scheduler.hpp"
#include "tiramisu/serialize/checkpoint.hpp"

using namespace tiramisu;
using namespace tiramisu::nn;
using namespace tiramisu::optim;
using namespace tiramisu::autograd;

struct CharVocab {
  std::unordered_map<char, int64_t> char_to_id;
  std::vector<char> id_to_char;

  int64_t size() const { return static_cast<int64_t>(id_to_char.size()); }

  void build(const std::string& text) {
    std::vector<char> chars;
    for (char c : text) {
      if (char_to_id.find(c) == char_to_id.end()) {
        char_to_id[c] = static_cast<int64_t>(chars.size());
        chars.push_back(c);
      }
    }
    id_to_char = std::move(chars);
  }

  std::vector<int64_t> encode(const std::string& text) const {
    std::vector<int64_t> ids;
    ids.reserve(text.size());
    for (char c : text) {
      ids.push_back(char_to_id.at(c));
    }
    return ids;
  }

  std::string decode(const std::vector<int64_t>& ids) const {
    std::string out;
    out.reserve(ids.size());
    for (int64_t id : ids) {
      out.push_back(id_to_char[static_cast<size_t>(id)]);
    }
    return out;
  }
};

struct TrainConfig {
  std::string data_path = "data/tiny_shakespeare.txt";
  std::string checkpoint_path;
  std::string preset = "tiny";
  int epochs = 3;
  int batch_size = 8;
  int seq_len = 64;
  float lr = 3e-4f;
  float weight_decay = 0.1f;
  float grad_clip = 1.0f;
  int eval_interval = 50;
  int checkpoint_interval = 200;
  int max_batches = -1;
  bool generate = true;
  bool generate_only = false;
  std::string prompt = "First Citizen:\n";
  int sample_chars = 400;
  float temperature = 0.8f;
  int sample_seed = 42;
  bool use_cuda = false;
};

std::string read_file(const std::string& path) {
  std::ifstream file(path);
  if (!file) {
    throw std::runtime_error("Cannot open: " + path);
  }
  std::ostringstream ss;
  ss << file.rdbuf();
  return ss.str();
}

GPTConfig gpt_config_for_preset(const std::string& preset, int64_t vocab_size,
                                int seq_len) {
  if (preset == "10m") {
    return GPTConfig{
        .vocab_size = vocab_size,
        .d_model = 384,
        .num_heads = 6,
        .num_layers = 6,
        .max_seq_len = seq_len,
        .tie_weights = true,
    };
  }
  if (preset == "2m") {
    return GPTConfig{
        .vocab_size = vocab_size,
        .d_model = 200,
        .num_heads = 4,
        .num_layers = 4,
        .max_seq_len = seq_len,
        .tie_weights = true,
    };
  }
  return GPTConfig{
      .vocab_size = vocab_size,
      .d_model = 64,
      .num_heads = 2,
      .num_layers = 2,
      .max_seq_len = seq_len,
      .tie_weights = true,
  };
}

void init_parameters(std::vector<Tensor*>& params, std::mt19937& gen) {
  for (Tensor* p : params) {
    std::vector<float> host(static_cast<size_t>(p->numel()));
    if (p->shape().size() == 2) {
      const int64_t fan_in = p->shape()[0];
      std::normal_distribution<float> dist(
          0.0f, 0.02f / std::sqrt(static_cast<float>(fan_in)));
      for (float& v : host) {
        v = dist(gen);
      }
    } else if (p->shape().size() == 1) {
      std::fill(host.begin(), host.end(), 0.0f);
    }
    if (p->device() == Device::CPU) {
      std::copy(host.begin(), host.end(), p->data<float>());
    } else {
      cuda_mem::copy_bytes(host.data(), p->data<float>(),
                           host.size() * sizeof(float), Device::CPU,
                           p->device());
    }
  }
}

Tensor make_batch_tensor(const std::vector<int64_t>& flat_ids, int64_t batch,
                         int64_t seq, Device device) {
  std::vector<float> host(static_cast<size_t>(batch * seq));
  for (int64_t i = 0; i < batch * seq; ++i) {
    host[static_cast<size_t>(i)] =
        static_cast<float>(flat_ids[static_cast<size_t>(i)]);
  }
  Tensor t({batch, seq}, DType::Float32, device);
  cuda_mem::copy_bytes(host.data(), t.data<float>(), host.size() * sizeof(float),
                       Device::CPU, device);
  return t;
}

float tensor_scalar(const Tensor& t) {
  if (t.device() == Device::CUDA) {
    return cuda_mem::read_scalar_f32(t);
  }
  return t.data<float>()[0];
}

float evaluate_loss(GPT& model, const std::vector<int64_t>& ids, int64_t seq_len,
                    int64_t batch_size, Device device) {
  const int64_t vocab = model.config().vocab_size;
  float total = 0.0f;
  int64_t batches = 0;

  for (size_t start = 0; start + static_cast<size_t>(seq_len) + 1 <= ids.size();
       start += static_cast<size_t>(batch_size * seq_len)) {
    std::vector<int64_t> input_ids;
    std::vector<int64_t> target_ids;
    input_ids.reserve(static_cast<size_t>(batch_size * seq_len));
    target_ids.reserve(static_cast<size_t>(batch_size * seq_len));

    for (int64_t b = 0; b < batch_size; ++b) {
      size_t offset = start + static_cast<size_t>(b * seq_len);
      if (offset + static_cast<size_t>(seq_len) + 1 > ids.size()) {
        break;
      }
      for (int64_t s = 0; s < seq_len; ++s) {
        input_ids.push_back(ids[offset + static_cast<size_t>(s)]);
        target_ids.push_back(ids[offset + static_cast<size_t>(s) + 1]);
      }
    }

    if (input_ids.empty()) break;
    const int64_t actual_batch = static_cast<int64_t>(input_ids.size()) / seq_len;

    Tensor batch_x = make_batch_tensor(input_ids, actual_batch, seq_len, device);
    Tensor batch_y = make_batch_tensor(target_ids, actual_batch, seq_len, device);

    Tensor logits = model.forward(batch_x);
    Tensor flat_logits =
        reshape(logits, {actual_batch * seq_len, vocab});
    Tensor flat_targets = batch_y.reshape({actual_batch * seq_len});
    Tensor loss = cross_entropy_loss(flat_logits, flat_targets);
    total += tensor_scalar(loss);
    batches++;
  }

  return batches > 0 ? total / static_cast<float>(batches) : 0.0f;
}

int64_t sample_next_token(const float* logits, int64_t vocab, float temperature,
                          std::mt19937& rng) {
  if (temperature <= 0.0f) {
    int64_t best = 0;
    for (int64_t i = 1; i < vocab; ++i) {
      if (logits[i] > logits[best]) {
        best = i;
      }
    }
    return best;
  }

  float max_logit = logits[0];
  for (int64_t i = 1; i < vocab; ++i) {
    if (logits[i] > max_logit) {
      max_logit = logits[i];
    }
  }

  std::vector<float> probs(static_cast<size_t>(vocab));
  float sum = 0.0f;
  for (int64_t i = 0; i < vocab; ++i) {
    probs[static_cast<size_t>(i)] =
        std::exp((logits[i] - max_logit) / temperature);
    sum += probs[static_cast<size_t>(i)];
  }

  std::uniform_real_distribution<float> dist(0.0f, sum);
  const float target = dist(rng);
  float cumulative = 0.0f;
  for (int64_t i = 0; i < vocab; ++i) {
    cumulative += probs[static_cast<size_t>(i)];
    if (target <= cumulative) {
      return i;
    }
  }
  return vocab - 1;
}

std::string generate_text(GPT& model, const CharVocab& vocab,
                          const std::string& prompt, int64_t max_new_tokens,
                          float temperature, std::mt19937& rng,
                          Device device) {
  NoGradGuard guard;
  std::vector<int64_t> context = vocab.encode(prompt);
  const size_t prompt_len = context.size();

  for (int64_t t = 0; t < max_new_tokens; ++t) {
    Tensor ids =
        make_batch_tensor(context, 1, static_cast<int64_t>(context.size()), device);
    Tensor logits = model.forward(ids);
    const int64_t vocab_size = model.config().vocab_size;
    const int64_t last = logits.shape()[1] - 1;

    std::vector<float> row(static_cast<size_t>(vocab_size));
    if (device == Device::CUDA) {
      cuda_mem::copy_bytes(
          logits.data<float>() + last * vocab_size, row.data(),
          static_cast<std::size_t>(vocab_size) * sizeof(float), Device::CUDA,
          Device::CPU);
    } else {
      std::copy_n(logits.data<float>() + last * vocab_size, vocab_size,
                  row.data());
    }

    const int64_t next_id =
        sample_next_token(row.data(), vocab_size, temperature, rng);
    context.push_back(next_id);
    if (context.size() > static_cast<size_t>(model.config().max_seq_len)) {
      context.erase(context.begin());
    }
  }

  std::vector<int64_t> generated(context.begin() + static_cast<long>(prompt_len),
                                 context.end());
  return vocab.decode(generated);
}

TrainConfig parse_args(int argc, char** argv) {
  TrainConfig cfg;
  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    auto next = [&]() -> std::string {
      if (i + 1 >= argc) throw std::runtime_error("missing value for " + arg);
      return argv[++i];
    };
    if (arg == "--data") {
      cfg.data_path = next();
    } else if (arg == "--preset") {
      cfg.preset = next();
    } else if (arg == "--epochs") {
      cfg.epochs = std::stoi(next());
    } else if (arg == "--batch-size") {
      cfg.batch_size = std::stoi(next());
    } else if (arg == "--seq-len") {
      cfg.seq_len = std::stoi(next());
    } else if (arg == "--lr") {
      cfg.lr = std::stof(next());
    } else if (arg == "--checkpoint") {
      cfg.checkpoint_path = next();
    } else if (arg == "--max-batches") {
      cfg.max_batches = std::stoi(next());
    } else if (arg == "--no-generate") {
      cfg.generate = false;
    } else if (arg == "--generate-only") {
      cfg.generate_only = true;
      cfg.generate = true;
      cfg.epochs = 0;
    } else if (arg == "--prompt") {
      cfg.prompt = next();
    } else if (arg == "--sample-chars") {
      cfg.sample_chars = std::stoi(next());
    } else if (arg == "--temperature") {
      cfg.temperature = std::stof(next());
    } else if (arg == "--sample-seed") {
      cfg.sample_seed = std::stoi(next());
    } else if (arg == "--cuda") {
      cfg.use_cuda = true;
    } else if (arg == "--help") {
      std::printf(
          "Usage: train_shakespeare [options]\n"
          "  --data PATH          Corpus file (default: data/tiny_shakespeare.txt)\n"
          "  --preset tiny|2m|10m Model size preset (default: tiny)\n"
          "  --epochs N           Training epochs (default: 3)\n"
          "  --batch-size N       Batch size (default: 8)\n"
          "  --seq-len N          Sequence length (default: 64, use 256 for 10m)\n"
          "  --lr F               Base learning rate (default: 3e-4)\n"
          "  --checkpoint PATH    Save/load checkpoint path\n"
          "  --max-batches N      Stop after N optimizer steps\n"
          "  --no-generate        Skip sampling after training\n"
          "  --generate-only      Load checkpoint and sample (implies --epochs 0)\n"
          "  --prompt TEXT        Generation prompt (default: \"First Citizen:\\n\")\n"
          "  --sample-chars N     Characters to generate (default: 400)\n"
          "  --temperature F      Sampling temperature, 0=greedy (default: 0.8)\n"
          "  --sample-seed N      RNG seed for sampling (default: 42)\n"
          "  --cuda               Train on GPU (requires CUDA build)\n");
      std::exit(0);
    }
  }
  if (cfg.preset == "10m" && cfg.seq_len == 64) {
    cfg.seq_len = 256;
    cfg.batch_size = 16;
    cfg.epochs = 5;
  }
  if (cfg.preset == "2m" && cfg.seq_len == 64) {
    cfg.seq_len = 128;
    cfg.batch_size = 16;
  }
  return cfg;
}

int main(int argc, char** argv) {
  const TrainConfig cfg = parse_args(argc, argv);

  Device device = Device::CPU;
  if (cfg.use_cuda) {
#ifdef TIRAMISU_CUDA_ENABLED
    if (!cuda_available()) {
      throw std::runtime_error("--cuda requested but no CUDA device is available");
    }
    set_device(Device::CUDA);
    device = Device::CUDA;
#else
    throw std::runtime_error("--cuda requested but tiramisu was built without CUDA");
#endif
  }

  const std::string text = read_file(cfg.data_path);
  CharVocab vocab;
  vocab.build(text);
  const std::vector<int64_t> ids = vocab.encode(text);

  const size_t split = ids.size() * 9 / 10;
  std::vector<int64_t> train_ids(ids.begin(), ids.begin() + static_cast<long>(split));
  std::vector<int64_t> val_ids(ids.begin() + static_cast<long>(split), ids.end());

  GPTConfig model_cfg =
      gpt_config_for_preset(cfg.preset, vocab.size(), cfg.seq_len);
  GPT model(gpt_config_for_preset(cfg.preset, vocab.size(), cfg.seq_len), device);
  std::vector<Tensor*> params = model.parameters();
  std::mt19937 gen(1234);
  init_parameters(params, gen);

  int64_t resume_step = 0;
  int64_t resume_epoch = 0;
  if (!cfg.checkpoint_path.empty()) {
    std::ifstream ck(cfg.checkpoint_path);
    if (ck.good()) {
      serialize::load_gpt_model(cfg.checkpoint_path, model, &resume_step,
                                &resume_epoch);
      std::printf("Loaded checkpoint from %s (step %lld, epoch %lld)\n",
                  cfg.checkpoint_path.c_str(),
                  static_cast<long long>(resume_step),
                  static_cast<long long>(resume_epoch));
    }
  }

  AdamW optimizer(params, cfg.lr, 0.9f, 0.999f, 1e-8f, cfg.weight_decay);

  const int64_t steps_per_epoch =
      std::max<int64_t>(1, static_cast<int64_t>(train_ids.size()) /
                               (cfg.batch_size * cfg.seq_len));
  const int64_t total_steps = steps_per_epoch * cfg.epochs;
  CosineAnnealingLR scheduler(cfg.lr, total_steps, cfg.lr * 0.1f);

  std::printf("Preset: %s | device=%s | vocab=%lld d_model=%lld layers=%lld heads=%lld\n",
              cfg.preset.c_str(), device == Device::CUDA ? "cuda" : "cpu",
              static_cast<long long>(model_cfg.vocab_size),
              static_cast<long long>(model_cfg.d_model),
              static_cast<long long>(model_cfg.num_layers),
              static_cast<long long>(model_cfg.num_heads));
  std::printf("Parameters: %lld | train tokens=%zu | val tokens=%zu\n",
              static_cast<long long>(model.count_parameters()), train_ids.size(),
              val_ids.size());

  int64_t global_step = resume_step;
  if (!cfg.generate_only) {
  for (int epoch = resume_epoch; epoch < cfg.epochs; ++epoch) {
    float epoch_loss = 0.0f;
    int64_t batch_count = 0;

    for (size_t start = 0;
         start + static_cast<size_t>(cfg.seq_len) + 1 <= train_ids.size();
         start += static_cast<size_t>(cfg.batch_size * cfg.seq_len)) {
      if (cfg.max_batches >= 0 && global_step >= cfg.max_batches) {
        break;
      }

      std::vector<int64_t> input_ids;
      std::vector<int64_t> target_ids;
      input_ids.reserve(static_cast<size_t>(cfg.batch_size * cfg.seq_len));
      target_ids.reserve(static_cast<size_t>(cfg.batch_size * cfg.seq_len));

      for (int b = 0; b < cfg.batch_size; ++b) {
        size_t offset = start + static_cast<size_t>(b * cfg.seq_len);
        if (offset + static_cast<size_t>(cfg.seq_len) + 1 > train_ids.size()) {
          break;
        }
        for (int s = 0; s < cfg.seq_len; ++s) {
          input_ids.push_back(train_ids[offset + static_cast<size_t>(s)]);
          target_ids.push_back(train_ids[offset + static_cast<size_t>(s) + 1]);
        }
      }
      if (input_ids.empty()) break;

      const int64_t actual_batch =
          static_cast<int64_t>(input_ids.size()) / cfg.seq_len;
      Tensor batch_x =
          make_batch_tensor(input_ids, actual_batch, cfg.seq_len, device);
      Tensor batch_y =
          make_batch_tensor(target_ids, actual_batch, cfg.seq_len, device);

      optimizer.zero_grad();
      Tensor logits = model.forward(batch_x);
      Tensor flat_logits =
          reshape(logits, {actual_batch * cfg.seq_len, model_cfg.vocab_size});
      Tensor flat_targets = batch_y.reshape({actual_batch * cfg.seq_len});
      Tensor loss = cross_entropy_loss(flat_logits, flat_targets);
      backward(loss);

      clip_grad_norm(params, cfg.grad_clip);
      optimizer.step();
      global_step++;

      const float lr = scheduler.step();
      optimizer.set_lr(lr);

      epoch_loss += tensor_scalar(loss);
      batch_count++;

      if (global_step % cfg.eval_interval == 0) {
        const float val_loss =
            evaluate_loss(model, val_ids, cfg.seq_len, cfg.batch_size, device);
        std::printf(
            "step %lld | train_loss %.4f | val_loss %.4f | lr %.6f\n",
            static_cast<long long>(global_step), tensor_scalar(loss), val_loss,
            lr);
      }

      if (!cfg.checkpoint_path.empty() &&
          global_step % cfg.checkpoint_interval == 0) {
        serialize::save_gpt_model(cfg.checkpoint_path, model, global_step, epoch);
        std::printf("Saved checkpoint to %s\n", cfg.checkpoint_path.c_str());
      }
    }

    const float avg_loss =
        batch_count > 0 ? epoch_loss / static_cast<float>(batch_count) : 0.0f;
    const float val_loss =
        evaluate_loss(model, val_ids, cfg.seq_len, cfg.batch_size, device);
    std::printf("epoch %d | avg_train_loss %.4f | val_loss %.4f\n", epoch + 1,
                avg_loss, val_loss);

    if (cfg.max_batches >= 0 && global_step >= cfg.max_batches) {
      break;
    }
  }
  }

  if (!cfg.generate_only && !cfg.checkpoint_path.empty()) {
    serialize::save_gpt_model(cfg.checkpoint_path, model, global_step,
                              cfg.epochs);
    std::printf("Final checkpoint saved to %s\n", cfg.checkpoint_path.c_str());
  }

  if (cfg.generate) {
    std::mt19937 sample_rng(cfg.sample_seed);
    const std::string sample = generate_text(
        model, vocab, cfg.prompt, cfg.sample_chars, cfg.temperature, sample_rng,
        device);
    std::printf("\n--- sample (prompt: %s) ---\n%s\n", cfg.prompt.c_str(),
                sample.c_str());
  }

  return 0;
}

File: examples/python/train_shakespeare.py
#!/usr/bin/env python3
"""Train a char-level GPT on Tiny Shakespeare using the pip-installable API.

Mirrors examples/train_shakespeare.cpp. Presets: tiny, 2m, 10m.
"""

from __future__ import annotations

import argparse
import math
import time
from pathlib import Path

import numpy as np
import tiramisu as tr


PRESETS = {
    "tiny": dict(d_model=64, num_heads=2, num_layers=2, seq_len=64, batch_size=8),
    "2m":   dict(d_model=200, num_heads=4, num_layers=4, seq_len=128, batch_size=16),
    "10m":  dict(d_model=384, num_heads=6, num_layers=6, seq_len=256, batch_size=16),
}


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser()
    p.add_argument("--data", default="data/tiny_shakespeare.txt")
    p.add_argument("--preset", choices=list(PRESETS), default="tiny")
    p.add_argument("--epochs", type=int, default=3)
    p.add_argument("--lr", type=float, default=3e-4)
    p.add_argument("--weight-decay", type=float, default=0.1)
    p.add_argument("--grad-clip", type=float, default=1.0)
    p.add_argument("--checkpoint", type=str, default=None)
    p.add_argument("--checkpoint-interval", type=int, default=200)
    p.add_argument("--eval-interval", type=int, default=50)
    p.add_argument("--max-batches", type=int, default=-1,
                   help="Cap total optimizer steps (for smoke tests).")
    p.add_argument("--resume", action="store_true",
                   help="With --checkpoint, load and continue training.")
    p.add_argument("--device", choices=["cpu", "cuda"], default="cpu",
                   help="Run model + tensors on cpu or cuda (requires CUDA build).")
    return p.parse_args()


def build_vocab(text: str) -> tuple[dict[str, int], list[str]]:
    chars: list[str] = []
    seen: dict[str, int] = {}
    for c in text:
        if c not in seen:
            seen[c] = len(chars)
            chars.append(c)
    return seen, chars


def encode(text: str, ctoi: dict[str, int]) -> np.ndarray:
    return np.fromiter((ctoi[c] for c in text), dtype=np.float32, count=len(text))


def iter_batches(ids: np.ndarray, batch_size: int, seq_len: int):
    """Yield (input, target) uint tensors of shape (B, S), sliding window over `ids`."""
    step = batch_size * seq_len
    n = len(ids)
    for start in range(0, n - seq_len - 1, step):
        rows_x, rows_y = [], []
        for b in range(batch_size):
            offset = start + b * seq_len
            if offset + seq_len + 1 > n:
                break
            rows_x.append(ids[offset:offset + seq_len])
            rows_y.append(ids[offset + 1:offset + seq_len + 1])
        if not rows_x:
            return
        yield np.stack(rows_x), np.stack(rows_y)


def evaluate(model: tr.nn.GPT, val_ids: np.ndarray, vocab_size: int,
             seq_len: int, batch_size: int, device: str,
             max_batches: int = 20) -> float:
    total, count = 0.0, 0
    for i, (bx, by) in enumerate(iter_batches(val_ids, batch_size, seq_len)):
        if i >= max_batches:
            break
        logits = model.forward(tr.from_numpy(bx, device=device))
        b, s = bx.shape
        flat = tr.reshape(logits, [b * s, vocab_size])
        targets = tr.from_numpy(by.reshape(-1), device=device)
        loss = tr.nn.cross_entropy_loss(flat, targets)
        total += float(loss.cpu().numpy()[0])
        count += 1
    return total / max(count, 1)


def main() -> None:
    args = parse_args()
    cfg = PRESETS[args.preset]

    text = Path(args.data).read_text()
    ctoi, itoc = build_vocab(text)
    vocab_size = len(itoc)
    ids = encode(text, ctoi)
    split = int(0.9 * len(ids))
    train_ids, val_ids = ids[:split], ids[split:]

    print(f"preset={args.preset} | vocab={vocab_size} | "
          f"train={len(train_ids):,} val={len(val_ids):,} tokens")

    if args.device == "cuda" and not tr.cuda_available():
        raise RuntimeError(
            "--device cuda requires a CUDA-enabled build. Reinstall with:\n"
            "  pip install --no-binary=tiramisu-ml \\\n"
            "    --config-settings=cmake.define.TIRAMISU_ENABLE_CUDA=ON \\\n"
            "    tiramisu-ml"
        )

    model = tr.nn.GPT(
        vocab_size=vocab_size,
        d_model=cfg["d_model"],
        num_heads=cfg["num_heads"],
        num_layers=cfg["num_layers"],
        max_seq_len=cfg["seq_len"],
        tie_weights=True,
        device=args.device,
    )
    print(f"device={args.device}")

    resume_step = 0
    resume_epoch = 0
    if args.checkpoint and args.resume:
        resume_step, resume_epoch = tr.serialize.load_gpt(args.checkpoint, model)
        print(f"resumed from step={resume_step} epoch={resume_epoch}")

    params = model.parameters()
    opt = tr.optim.AdamW(params, lr=args.lr, weight_decay=args.weight_decay)
    opt.step_count = resume_step

    steps_per_epoch = max(1, len(train_ids) // (cfg["batch_size"] * cfg["seq_len"]))
    scheduler = tr.optim.CosineAnnealingLR(
        base_lr=args.lr,
        total_steps=steps_per_epoch * args.epochs,
        min_lr=args.lr * 0.1,
    )

    global_step = resume_step
    start_epoch = resume_epoch if args.resume else 0
    target_epoch = (resume_epoch + args.epochs) if args.resume else args.epochs

    t0 = time.time()
    for epoch in range(start_epoch, target_epoch):
        epoch_loss = 0.0
        batches = 0
        for bx, by in iter_batches(train_ids, cfg["batch_size"], cfg["seq_len"]):
            if 0 <= args.max_batches <= global_step:
                break

            opt.zero_grad()
            logits = model.forward(tr.from_numpy(bx, device=args.device))
            b, s = bx.shape
            flat_logits = tr.reshape(logits, [b * s, vocab_size])
            flat_targets = tr.from_numpy(by.reshape(-1), device=args.device)
            loss = tr.nn.cross_entropy_loss(flat_logits, flat_targets)
            loss.backward()

            tr.optim.clip_grad_norm_(params, args.grad_clip)
            opt.step()
            global_step += 1

            lr = scheduler.step()
            opt.lr = lr

            train_loss = float(loss.cpu().numpy()[0])
            epoch_loss += train_loss
            batches += 1

            if global_step % args.eval_interval == 0:
                val = evaluate(model, val_ids, vocab_size,
                               cfg["seq_len"], cfg["batch_size"], args.device)
                dt = time.time() - t0
                print(f"step {global_step:>5} | train {train_loss:.4f} | "
                      f"val {val:.4f} | lr {lr:.6f} | {dt:.1f}s")

            if args.checkpoint and global_step % args.checkpoint_interval == 0:
                tr.serialize.save_gpt(args.checkpoint, model, global_step, epoch)

        avg = epoch_loss / max(batches, 1)
        print(f"epoch {epoch}: avg_loss={avg:.4f}")
        if args.checkpoint:
            tr.serialize.save_gpt(args.checkpoint, model, global_step, epoch + 1)


if __name__ == "__main__":
    main()
